telecine/src/main/java/com/jakewharton/telecine/RecordingSession.java

```diff
@@ -156,17 +158,53 @@ private void startRecording() {
     int sizePercentage = videoSizePercentage.get();
     Timber.d("Video size: %s%%", sizePercentage);
 
-    DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
-    int displayHeight = displayMetrics.heightPixels * sizePercentage / 100;
-    int displayWidth = displayMetrics.widthPixels * sizePercentage / 100;
-    int displayDpi = displayMetrics.densityDpi;
+    DisplayMetrics displayMetrics = new DisplayMetrics();
+    WindowManager wm = (WindowManager) context.getSystemService(WINDOW_SERVICE);
+    wm.getDefaultDisplay().getRealMetrics(displayMetrics);
+
+    Configuration configuration = context.getResources().getConfiguration();
+
+    int maximumWidth;
+    int maximumHeight;
+
+    // Get the best camera profile available. We assume MediaRecorder supports the highest.
+    CamcorderProfile camcorderProfile = CamcorderProfile.get(CamcorderProfile.QUALITY_HIGH);
+    if (camcorderProfile != null) {
+      boolean isLandscape = configuration.orientation == Configuration.ORIENTATION_LANDSCAPE;
+      int frameWidth = camcorderProfile.videoFrameWidth;
+      int frameHeight = camcorderProfile.videoFrameHeight;
+      maximumWidth = isLandscape ? frameWidth : frameHeight;
+      maximumHeight = isLandscape ? frameHeight : frameWidth;
+
+      // Check if video recording quality is higher than the screen resolution
+      if(maximumWidth > displayMetrics.widthPixels && maximumHeight > displayMetrics.heightPixels) {
+        maximumWidth = displayMetrics.widthPixels;
+        maximumHeight = displayMetrics.heightPixels;
+      } else {
+        // If not then we calculate new width or height to preserve aspect ratio
+        if(isLandscape) {
+          maximumWidth = displayMetrics.widthPixels * maximumHeight / displayMetrics.heightPixels;
```

Comment:
Well, in pratice, it's the width that is always calculated, because the maximum height quality of the camera will be one of: `480p`, `720p`, `1080p` or `2160p`. The `isLandscape` is just to check the orientation we're recording.

For example, with my device Nexus 7 which is taller (1200p) than the camera's capability (1080p) would have a new width of 1728p, resulting in aspecto ratio of `8:5`